Feat/genre search support

jozefKruszynski opened the pull request:
Initial draft PR in order to get some feedback on progress thus far.

jozefKruszynski committed:
style(genre): mypy and  ruff fixes

jozefKruszynski committed:
fix(genre): add auth to genre crud ops

marcelveldt commented on music_assistant/controllers/media/albums.py:
```diff
@@ -111,6 +111,7 @@ async def library_items(
         provider: str | list[str] | None = None,
         extra_query: str | None = None,
         extra_query_params: dict[str, Any] | None = None,
+        genre_items_ids: list[int] | None = None,
```
I would just name this parameter genres and then explain in the docstring that you want library genre id's here

marcelveldt commented on music_assistant/controllers/media/base.py:
```diff
@@ -799,8 +828,17 @@ def _apply_filters(
+            query_parts.append(
+                f"({self.db_table}.search_name LIKE :search OR "
+                f"{self.db_table}.item_id IN ({ids_str}))"
+            )
```
This seems incorrect to me

marcelveldt commented on music_assistant/controllers/media/base.py:
```diff
@@ -799,8 +828,17 @@ def _apply_filters(
             query_parts.append(f"{self.db_table}.search_name LIKE :search")
+        elif genre_items_ids:
+            ids_str = ",".join(str(x) for x in genre_items_ids)
+            query_parts.append(f"{self.db_table}.item_id IN ({ids_str})")
```
this is also incorrect - you are modifying the base class here

jozefKruszynski committed:
refactor(genre): naming and docstrings

jozefKruszynski committed:
refactor(genre): remove unnecessary base modifications

marcelveldt commented on music_assistant/controllers/media/genres.py:
```diff
@@ -1,60 +1,691 @@
+        # we debounce this to prevent too many rebuilds
+        self.mass.call_later(10, self.rebuild_index, task_id="rebuild_genre_index")
+
+    async def rebuild_index(self) -> None:
```
I have no idea why this is needed. Cant we do this with pure db queries ?
You probably just need a genre_mappings tables

marcelveldt commented on music_assistant/controllers/music.py:
```diff
@@ -2117,6 +2195,38 @@ async def __create_database_tables(self) -> None:
+            [artist_ids] json DEFAULT '[]',
+            [playlist_ids] json DEFAULT '[]',
+            [podcast_ids] json DEFAULT '[]',
+            [audiobook_ids] json DEFAULT '[]'
```
So this is the part I have doubts about - I think we can do better than this by having a genre_mappings table. So basically the same as how provider_mappings work currently

marcelveldt commented on music_assistant/controllers/music.py:
```diff
@@ -2117,6 +2195,38 @@ async def __create_database_tables(self) -> None:
+            f"""
+            CREATE TABLE IF NOT EXISTS {DB_TABLE_GENRE_ALIASES}(
+            [genre_id] INTEGER NOT NULL,
+            [alias] TEXT NOT NULL,
```
I would investigate storing aliases as json column on the genres table. 
sqlite has a native json handler nowadays which can also perform fast queries on the json and if I'm not mistaken can do indexes as well.

jozefKruszynski replied:
Gotcha, sounds far more sensible. I will probably be able to do a bit on Friday morning, but after that I'm away in the UK until Late Monday night.

jozefKruszynski committed:
refactor(genre): update to use json mapping column, rebase onto dev

Updated to use a json mapping column in the same way as provider
mappings, reabsed onto dev, image updates on genres fixed

jozefKruszynski committed:
style: added comment regarding metadata updates

jozefKruszynski committed:
fix: cleanup

MarvinSchenkel commented on music_assistant/controllers/media/genres.py:
```diff
@@ -1,60 +1,747 @@
+
+        self.logger.info("Finished background genre index scan. Updated %s genres.", count)
+
+    async def _scan_media_type(
```
I think this logic (along with the rebuild_index function) should not be in the controller. I think it makes more sense to create a proper database design for mappings between genres and media_items, rather than trying to constantly sync them off the metadata.genre attributes. The logic in these functions should in my opinion be a migration script that is run only once in order to migrate the existing genre metadata into the new database structure. Then we just need to update our other controllers to insert directly into the genre_mappings rather than writing to metadata.

So the database could look some thing like this
`[track] / [album] etc..` 1 -- * `[media_item_genre_mappings]` * -- 1 `[genre]`

Thoughts?

jozefKruszynski replied:
Sounds like a reasonable suggestion, it makes genre a first class citizen rather than an after thought and I think that thinking about genre in such a way, is actually essential to making something meaningful.

jozefKruszynski committed:
fix(genre): changes before rebase

jozefKruszynski committed:
more changes after rebasing
